### PR TITLE
Fanout metric

### PR DESCRIPTION
**Pull request by Vitaly-Protasov**
Solved #450

### CODE REVIEW

**Comment by 0crat** [comment]
This pull request [#451](https://github.com/yegor256/aibolit/issues/451) is assigned to @aravij[/z](https://www.0crat.com/u/aravij), here is [why](https://www.0crat.com/footprint/CT5UB7XAP/638e5de0-0dfc-4f0c-8b50-8c51e511316d); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @acheshkov[/z](https://www.0crat.com/u/acheshkov) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

**Comment by Vitaly-Protasov** [comment]
> 1. There can't be 2 classes in 1 Java file
> 2. It should show the number of used class excepting standard Java library.
> 3. You must check all imports, since you can redefine you own class.
>    E.g., _BufferedReader_ can be your own class not, stand Java library, it depends on the import
> 4. It should show the number of other classes referenced by a class. It is a metric, why are you showing the lines of code?
>    All test cases are completely wrong, need to discuss
> 5. You should check inner interfaces, inner classes, inner Anonymous classes

According to the our controversy, resolved smth from that

File: aibolit/metrics/fanout/FanOut.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            new_class = tree.get_attr(name_node, 'string').split('.')[-1]
+            if considered_classes.get(new_class) is None:
+                fan_outs += 1
+                considered_classes[new_class] = 0
+
+        for i in tree.subtrees_with_root_type(ASTNodeType.CLASS_DECLARATION):
+            ast = AST(tree.tree.subgraph(i), i[0])
+            for j in ast.subtrees_with_root_type(ASTNodeType.VARIABLE_DECLARATOR):
+                ast_ = AST(tree.tree.subgraph(j), j[0])
+                for inv_class in ast_.subtrees_with_root_type(ASTNodeType.CLASS_CREATOR):
```
Comment (aravij) [inline aibolit/metrics/fanout/FanOut.py:44]: `inv_class` sounds like `inverse_class`. We better use here more proper name.

File: aibolit/metrics/fanout/FanOut.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                    name_node, = islice(ast_.children_with_type(j[0], ASTNodeType.STRING), 1)
+                    new_class_name = ast_.get_attr(name_node, 'string')
+                    ast__ = AST(tree.tree.subgraph(inv_class), inv_class[0])
+
+                    for each_name in ast__.subtrees_with_root_type(ASTNodeType.REFERENCE_TYPE):
+                        name_node, = islice(ast__.children_with_type(each_name[0], ASTNodeType.STRING), 1)
+                        used_class_name = ast__.get_attr(name_node, 'string')
+
+                        if considered_classes.get(used_class_name) is None:
```
Comment (aravij) [inline aibolit/metrics/fanout/FanOut.py:53]:
```suggestion
                        if used_class_name not in considered_classes:
```

File: aibolit/metrics/fanout/FanOut.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+                            considered_classes[used_class_name] = 0
+                            fan_outs += 1
+                        if considered_classes.get(new_class_name) is None:
```
Comment (aravij) [inline aibolit/metrics/fanout/FanOut.py:56]:
```suggestion
                        if new_class_name not in considered_classes:
```

File: aibolit/metrics/fanout/FanOut.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+                            considered_classes[new_class_name] = 0
+
+            # check classes of invokated methods
+            for j in ast.subtrees_with_root_type(ASTNodeType.STATEMENT_EXPRESSION):
+                ast1 = AST(tree.tree.subgraph(j), j[0])
+                for each_method in ast1.subtrees_with_root_type(ASTNodeType.METHOD_INVOCATION):
+                    name_of_invoked_class = ast1.get_method_invocation_params(each_method[0])
+                    if considered_classes.get(name_of_invoked_class.object_name) is None:
```
Comment (aravij) [inline aibolit/metrics/fanout/FanOut.py:64]:
```suggestion
                    if name_of_invoked_class.object_name not in considered_classes:
```

File: aibolit/metrics/fanout/FanOut.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+                              'int': 0, 'long': 0,
+                              }
+        fan_outs = 0
+
+        # check imported classes
+        tree = AST.build_from_javalang(build_ast(filename))
+        for each_import in (tree.children_with_type(tree.root, ASTNodeType.IMPORT)):
+            name_node, = islice(tree.children_with_type(each_import, ASTNodeType.STRING), 1)
+            new_class = tree.get_attr(name_node, 'string').split('.')[-1]
+            if considered_classes.get(new_class) is None:
```
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply aibolit/metrics/fanout/FanOut.py:37]: Didn't change

File: aibolit/metrics/fanout/FanOut.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+                fan_outs += 1
+                considered_classes[new_class] = 0
+
+        for i in tree.subtrees_with_root_type(ASTNodeType.CLASS_DECLARATION):
```
[the remark this answers is not in the document]
Comment (Vitaly-Protasov) [reply aibolit/metrics/fanout/FanOut.py]: Fixed

File: aibolit/metrics/fanout/FanOut.py (hunk 1 above)
Comment (Vitaly-Protasov) [reply aibolit/metrics/fanout/FanOut.py]: Rewrited

File: aibolit/metrics/fanout/FanOut.py (hunk 2 above)
Comment (Vitaly-Protasov) [reply aibolit/metrics/fanout/FanOut.py:52]: Didn't change

File: aibolit/metrics/fanout/FanOut.py (hunk 3 above)
Comment (Vitaly-Protasov) [reply aibolit/metrics/fanout/FanOut.py:55]: Didn't change

File: aibolit/metrics/fanout/FanOut.py (hunk 4 above)
Comment (Vitaly-Protasov) [reply aibolit/metrics/fanout/FanOut.py]: Didn't change

**Comment by 0crat** [comment]
Job was finished in 26 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Job audit: `ARC` didn't merge or close PR, performer @aravij[/z](https://www.0crat.com/u/aravij)[/z](https://www.0crat.com/u/aravij) didn't make CR comments

**Pull request merged**